In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load the data
df = pd.read_csv('Leads.csv')

In [3]:
# 1. See the size of the data
print(f"Dataset has {df.shape[0]} rows and {df.shape[1]} columns.\n")

Dataset has 9240 rows and 37 columns.



In [4]:
# 2. Check for missing values
missing_counts = df.isnull().sum()
missing_pct = (missing_counts / len(df)) * 100

In [9]:
# Combine into a summary table
missing_df = pd.DataFrame({
    'Missing Count' : missing_counts,
    'Percentage (%)' : missing_pct
}).sort_values(by='Percentage (%)', ascending=False)

print("Top 10 Columns with Missing Values:")
print(missing_df.head(15))

Top 10 Columns with Missing Values:
                                               Missing Count  Percentage (%)
Lead Quality                                            4767       51.590909
Asymmetrique Profile Index                              4218       45.649351
Asymmetrique Activity Score                             4218       45.649351
Asymmetrique Activity Index                             4218       45.649351
Asymmetrique Profile Score                              4218       45.649351
Tags                                                    3353       36.287879
What matters most to you in choosing a course           2709       29.318182
Lead Profile                                            2709       29.318182
What is your current occupation                         2690       29.112554
Country                                                 2461       26.634199
How did you hear about X Education                      2207       23.885281
Specialization                          

# Missing Columns Handling

The Asymmetrique Columns (~45.6% missing): These are proprietary lead scores generated by an external agency. Because almost half the data is missing and we don't know the exact logic behind how the agency calculated these scores, keeping them might introduce bias. Decision: Drop them.

Lead Quality (~51.5% missing): This is based on the sales employee's intuition. It's highly missing because many leads are new and haven't been evaluated yet. However, "Not Sure" or missing is an important data point itself. Instead of dropping it, we can create a new category called "Not Specified".

Tags (~36.3% missing): These are tags assigned by sales sub-teams. Like Lead Quality, we can impute the missing values with "Not Specified".

In [11]:
# 1. Drop the Asymmetrique columns completely
columns_to_drop = [
    'Asymmetrique Profile Index', 
    'Asymmetrique Activity Score', 
    'Asymmetrique Activity Index', 
    'Asymmetrique Profile Score'
    ]

df = df.drop(columns=columns_to_drop)

In [12]:
# 2. Impute high-missing categorical columns with a new category
columns_to_impute = ['Lead Quality', 'Tags', 'Lead Profile', 'What is your current occupation']
for col in columns_to_impute:
    df[col] = df[col].fillna('Not Specified')

In [13]:
# 3. Check the remaining missing values to see our progress
remaining_missing = df.isnull().sum().sort_values(ascending=False)
print("Remaining missing values (Top 5):")
print(remaining_missing.head(5))

Remaining missing values (Top 5):
What matters most to you in choosing a course    2709
Country                                          2461
How did you hear about X Education               2207
Specialization                                   1438
City                                             1420
dtype: int64
